### ISSUE TITLE
Provide pre-allocated output vector/matrix

### ISSUE THREAD

**Issue by jjaraalm** [autodiff/autodiff#182]
I have an application where I am routinely computing large Jacobin matrices. This is coupled to another library through a plain double array, and I end up having to manually copy everything between the output of jacobian and the array. Would it be possible to optionally give jacobian a pre-allocated output storage location directly to avoid copies? Being able to directly provide an `Eigen::Map<Eigen::MatrixXd>` would be ideal.

**Comment by allanleal** [comment]
In `autodiff/forward/utils/gradient.hpp`, we would need a `jacobian` function as follows:

~~~c++
template<typename Fun, typename... Vars, typename... Args, typename Y, typename Jac>
auto jacobian(const Fun& f, const Wrt<Vars...>& wrt, const At<Args...>& at, Y& F, Jac& J)
~~~

Would you be able to provide a PR with this change?

Then, the implementation of the existing `jacobian`:

~~~c++
template<typename Fun, typename... Vars, typename... Args, typename Y, typename Jac>
auto jacobian(const Fun& f, const Wrt<Vars...>& wrt, const At<Args...>& at, Y& F)
~~~

would use the previous one.

**Comment by jjaraalm** [comment]
Sure, I can work on that and put in a PR.

**Comment by jjaraalm** [comment]
Closed by #184
